consensus/src/epoch_manager.rs

```diff
@@ -263,9 +251,48 @@ impl EpochManager {
                     vec![1; proposers.len()]
                 };
 
+                // First two epochs are immediatelly skipped after a few transactions, so we don't want to look at their history.
+                let epoch_to_proposers = if use_history_from_previous_epoch && epoch_state.epoch > 2
+                {
+                    self.storage
+                        .aptos_db()
+                        .get_epoch_ending_ledger_infos(epoch_state.epoch - 2, epoch_state.epoch)
+                        .and_then(|proof| {
+                            ensure!(proof.ledger_info_with_sigs.len() == 2);
+                            let prev_epoch_state = proof.ledger_info_with_sigs[0]
+                                .ledger_info()
+                                .next_epoch_state()
+                                .ok_or_else(|| anyhow::anyhow!("no prev_epoch_state"))?;
+                            let cur_epoch_state = proof.ledger_info_with_sigs[1]
+                                .ledger_info()
+                                .next_epoch_state()
+                                .ok_or_else(|| anyhow::anyhow!("no cur_epoch_state"))?;
+                            ensure!(cur_epoch_state.epoch == epoch_state.epoch, "fetched epoch_ending ledger_infos are for a wrong epoch {} vs {}", cur_epoch_state.epoch, epoch_state.epoch);
+                            let prev_epoch_proposers = prev_epoch_state
+                                .verifier
+                                .get_ordered_account_addresses_iter()
+                                .collect::<Vec<_>>();
+                            let cur_epoch_proposers = cur_epoch_state
+                                .verifier
+                                .get_ordered_account_addresses_iter()
+                                .collect::<Vec<_>>();
+                            ensure!(proposers == cur_epoch_proposers, "proposers from state and fetched epoch_ending ledger_infos are missaligned");
```

Comment:
this block reads redundant, probably can have a lambda that does ledger_info -> (epoch, proposers) transformation?